server/sonar-server/src/main/java/org/sonar/server/plugins/ServerPluginRepository.java

```diff
@@ -19,76 +19,385 @@
+      deleteQuietly(info.getNonNullJarFile());
+      return;
+    }
+    PluginInfo existing = pluginInfosByKeys.put(info.getKey(), info);
+    if (existing != null) {
+      throw MessageException.of(format("Found two files for the same plugin [%s]: %s and %s",
+        info.getKey(), info.getNonNullJarFile().getName(), existing.getNonNullJarFile().getName()));
+    }
+
+  }
+
+  /**
+   * Move or copy plugin to directory extensions/plugins. If a version of this plugin
+   * already exists then it's deleted.
+   */
+  private void overrideAndRegisterPlugin(File sourceFile, boolean deleteSource) {
+    File destDir = fs.getInstalledPluginsDir();
+    File destFile = new File(destDir, sourceFile.getName());
+    if (destFile.exists()) {
+      // plugin with same filename already installed
+      deleteQuietly(destFile);
+    }
+
+    try {
+      if (deleteSource) {
+        moveFile(sourceFile, destFile);
+      } else {
+        copyFile(sourceFile, destFile, true);
+      }
+    } catch (IOException e) {
+      throw new IllegalStateException(format("Fail to move or copy plugin: %s to %s",
+        sourceFile.getAbsolutePath(), destFile.getAbsolutePath()), e);
+    }
+
+    PluginInfo info = PluginInfo.create(destFile);
+    PluginInfo existing = pluginInfosByKeys.put(info.getKey(), info);
+    if (existing != null) {
+      if (!existing.getNonNullJarFile().getName().equals(destFile.getName())) {
+        deleteQuietly(existing.getNonNullJarFile());
+      }
+      LOG.info("Plugin {} [{}] updated to version {}", info.getName(), info.getKey(), info.getVersion());
+    } else {
+      LOG.info("Plugin {} [{}] installed", info.getName(), info.getKey());
+    }
+  }
+
+  private void loadCorePlugins() {
+    for (File file : listJarFiles(fs.getCorePluginsDir())) {
+      PluginInfo info = PluginInfo.create(file).setCore(true);
+      registerPluginInfo(info);
+    }
+  }
+
+  /**
+   * Removes the plugins that are not compatible with current environment.
+   */
+  private void unloadIncompatiblePlugins() {
+    // loop as long as the previous loop ignored some plugins. That allows to support dependencies
+    // on many levels, for example D extends C, which extends B, which requires A. If A is not installed,
+    // then B, C and D must be ignored. That's not possible to achieve this algorithm with a single
+    // iteration over plugins.
+    Set<String> removedKeys = new HashSet<>();
+    do {
+      removedKeys.clear();
+      for (PluginInfo plugin : pluginInfosByKeys.values()) {
+        if (!isCompatible(plugin, server, pluginInfosByKeys)) {
+          removedKeys.add(plugin.getKey());
+        }
+      }
+      for (String removedKey : removedKeys) {
+        pluginInfosByKeys.remove(removedKey);
+      }
+    } while (!removedKeys.isEmpty());
+  }
+
+  @VisibleForTesting
+  static boolean isCompatible(PluginInfo plugin, Server server, Map<String, PluginInfo> allPluginsByKeys) {
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)The Cyclomatic Complexity of this method "isCompatible" is 14 which is greater than 10 authorized.[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)